cli/src/cmd/cast/bind.rs

```diff
@@ -0,0 +1,158 @@
+        default_value = DEFAULT_CRATE_VERSION,
+        value_name = "NAME"
+    )]
+    crate_version: String,
+
+    #[clap(flatten)]
+    chain: ClapChain,
+}
+
+impl Cmd for BindArgs {
+    type Output = BoxFuture<'static, eyre::Result<()>>;
+
+    fn run(self) -> eyre::Result<Self::Output> {
+        let cmd = Box::pin(async move {
+            let BindArgs {
+                path_or_address,
+                name,
+                etherscan_api_key,
+                bindings,
+                crate_name,
+                crate_version,
+                chain,
+            } = self.clone();
+
+            if Path::new(&path_or_address).exists() {
+                return self
+                    .generate_bindings(BindPath::Local {
+                        path: path_or_address,
+                        name: Some(name.unwrap_or_else(|| "Sample".to_owned())),
+                    })
+                    .await
+            } else {
+                self.generate_bindings(BindPath::Etherscan {
+                    address: path_or_address.parse::<Address>().unwrap(),
+                    chain: chain.inner,
+                    api_key: etherscan_api_key.unwrap(),
+                })
+                .await
+            }
+        });
+
+        Ok(cmd)
+    }
+}
+
+impl BindArgs {
+    pub async fn generate_bindings(self, address_or_path: BindPath) -> eyre::Result<()> {
+        match address_or_path {
+            BindPath::Etherscan { address, chain, api_key } => {
+                let client = Client::new(chain, api_key)?;
+                // get the source
+                let source = match client.contract_source_code(address).await {
+                    Ok(source) => source,
+                    Err(EtherscanError::InvalidApiKey) => {
+                        eyre::bail!("Invalid Etherscan API key. Did you set it correctly? You may be using an API key for another Etherscan API chain (e.g. Etherscan API key for Polygonscan).")
+                    }
+                    Err(EtherscanError::ContractCodeNotVerified(address)) => {
+                        eyre::bail!("Contract source code at {:?} on {} not verified. Maybe you have selected the wrong chain?", address, chain)
+                    }
+                    Err(err) => {
+                        eyre::bail!(err)
+                    }
+                };
+                let abigens = source
+                    .items
+                    .iter()
+                    .map(|item| Abigen::new(&item.contract_name.clone(), item.abi.clone()).unwrap())
+                    .collect::<Vec<Abigen>>();
+
+                let multi = MultiAbigen::from_abigens(abigens);
+
+                let bindings = multi.build().unwrap();
+                bindings.write_to_crate(
+                    &self.crate_name,
+                    &self.crate_version,
+                    self.get_binding_root(),
+                    false,
+                )?;
```

Comment:
Are we sure we want this API? I would think that maybe you want to generate a single-file binding, instead of having it codegen a full crate?

Reply:
Should we take it as input from the user or, by default, generate a single-file binding?

Reply:
your call, what is your expected UX? do you want bindings for every contract related with an address, or usually just a contract?

Reply:
If multiple ABI files are provided, then bindings for each one of them.  Bindings in separate files will only be there if the user asks explicitly otherwise bindings would be dumped in a single file(**lib.rs**).

Reply:
My 2c, but imho its important to also have the option to have it as a single file sans crate/module